tests/commands/test_store_client.py

```diff
@@ -0,0 +1,321 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+import json
+import logging
+import os
+from http.cookiejar import MozillaCookieJar, Cookie
+from unittest.mock import patch
+
+import pytest
+from macaroonbakery import httpbakery
+
+from charmcraft.cmdbase import CommandError
+from charmcraft.commands.store.client import Client, BASE_URL, _AuthHolder, visit_page_with_browser
+
+
+# --- AuthHolder tests
+
+@pytest.fixture
+def auth_holder(tmp_path):
+    """Produce an _AuthHolder instance with encapsulated resources.
+
+    In detail:
+
+    - writes into a temp dir
+
+    - with a mocked webbrowser module
+    """
+
+    ah = _AuthHolder()
+    ah._cookiejar_filepath = str(tmp_path / 'test.credentials')
+
+    with patch('webbrowser.open'):
+        yield ah
+
+
+def get_cookie(value='test'):
+    """Helper to create a cookie, which is quite long."""
+    return Cookie(
+        version=0, name='test-macaroon', value=value, port=None, port_specified=False,
+        domain='snapcraft.io', domain_specified=True, domain_initial_dot=False, path='/',
+        path_specified=True, secure=True, expires=2595425286, discard=False, comment=None,
+        comment_url=None, rest=None, rfc2109=False)
+
+
+def test_authholder_cookiejar_filepath():
+    """Builds the cookiejar filepath in user's config directory."""
+    with patch('appdirs.user_config_dir') as mock:
+        mock.return_value = 'testpath'
+        ah = _AuthHolder()
+
+    assert ah._cookiejar_filepath == 'testpath'
+    assert mock.called_once_with('charmcraft.credentials')
+
+
+def test_authholder_clear_credentials_ok(auth_holder, caplog):
+    """Clear credentials removes the file."""
+    caplog.set_level(logging.DEBUG, logger="charmcraft.commands")
+
+    with open(auth_holder._cookiejar_filepath, 'wb') as fh:
+        fh.write(b'fake credentials')
+
+    auth_holder.clear_credentials()
+
+    assert not os.path.exists(auth_holder._cookiejar_filepath)
+    expected = "Credentials cleared: file {!r} removed".format(auth_holder._cookiejar_filepath)
```

Comment:
is !r what we want? IIRC chipaca mentioned that it is close but not always correct

Reply:
repr() will also have some (rare) corners cases where would be ambiguous, but at least it's not confusing when spaces are used (which is a lot more common)